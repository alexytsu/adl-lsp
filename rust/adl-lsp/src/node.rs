#![allow(dead_code)]
use tree_sitter::Node;

pub enum NodeKind {
    // Comments and Documentation
    Comment,
    Docstring,

    // Identifiers and Names
    Identifier,
    ScopedName,
    TypeName,

    // Type System
    TypeParameters,
    TypeArguments,
    TypeExpression,
    PrimitiveType,

    // Type Definitions
    TypeDefinition,
    NewtypeDefinition,
    StructDefinition,
    UnionDefinition,
    Field,
    FieldBlock,

    // Module System
    Module,
    ModuleDefinition,
    ModuleBody,
    ImportDeclaration,
    ImportPath,

    // Annotations
    Annotation,
    Annotations,
    AnnotationDeclaration,

    // JSON
    JsonValue,
    JsonNumber,
    JsonString,
    JsonArray,
    JsonObject,
    JsonObjectPair,

    // Error
    Error,
}

impl NodeKind {
    pub fn is_user_defined_name(n: &Node) -> bool {
        Self::is_identifier(n) || Self::is_type_name(n) || Self::is_scoped_name(n)
    }

    pub fn is_definition(n: &Node) -> bool {
        Self::is_import_declaration(n) || Self::is_type_name(n) || Self::is_import_declaration(n)
    }

    /// Check if an identifier is part of a scoped name
    pub fn has_scoped_name_parent(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_scoped_name(&p))
    }

    /// Check if a node represents a valid identifier definition that can be referenced
    /// (i.e., an identifier that is part of a type_name)
    pub fn can_be_referenced(n: &Node) -> bool {
        Self::is_identifier(n) && n.parent().is_some_and(|p| Self::is_type_name(&p))
    }
}

impl NodeKind {
    pub fn from_str(s: &str) -> Self {
        match s {
            // Comments and Documentation
            "comment" => Self::Comment,
            "docstring" => Self::Docstring,

            // Identifiers and Names
            "identifier" => Self::Identifier,
            "scoped_name" => Self::ScopedName,
            "type_name" => Self::TypeName,

            // Type System
            "type_parameters" => Self::TypeParameters,
            "type_arguments" => Self::TypeArguments,
            "type_expression" => Self::TypeExpression,
            "primitive_type" => Self::PrimitiveType,

            // Type Definitions
            "type_definition" => Self::TypeDefinition,
            "newtype_definition" => Self::NewtypeDefinition,
            "struct_definition" => Self::StructDefinition,
            "union_definition" => Self::UnionDefinition,
            "field" => Self::Field,
            "field_block" => Self::FieldBlock,

            // Module System
            "module" => Self::Module,
            "module_definition" => Self::ModuleDefinition,
            "module_body" => Self::ModuleBody,
            "import_declaration" => Self::ImportDeclaration,
            "import_path" => Self::ImportPath,

            // Annotations
            "annotation" => Self::Annotation,
            "annotations" => Self::Annotations,
            "annotation_declaration" => Self::AnnotationDeclaration,

            // JSON
            "json_value" => Self::JsonValue,
            "json_number" => Self::JsonNumber,
            "json_string" => Self::JsonString,
            "json_array" => Self::JsonArray,
            "json_object" => Self::JsonObject,
            "json_object_pair" => Self::JsonObjectPair,

            // Default to Error for unknown kinds
            _ => Self::Error,
        }
    }

    pub fn as_str(&self) -> &str {
        match self {
            // Comments and Documentation
            Self::Comment => "comment",
            Self::Docstring => "docstring",

            // Identifiers and Names
            Self::Identifier => "identifier",
            Self::ScopedName => "scoped_name",

            // Type System
            Self::TypeName => "type_name",
            Self::TypeParameters => "type_parameters",
            Self::TypeExpression => "type_expression",
            Self::TypeArguments => "type_arguments",
            Self::PrimitiveType => "primitive_type",

            // Structs
            Self::StructDefinition => "struct_definition",
            Self::Field => "field",
            Self::FieldBlock => "field_block",

            // Unions
            Self::UnionDefinition => "union_definition",

            // Type Definitions
            Self::TypeDefinition => "type_definition",
            Self::NewtypeDefinition => "newtype_definition",

            // Module System
            Self::Module => "module",
            Self::ModuleDefinition => "module_definition",
            Self::ModuleBody => "module_body",
            Self::ImportDeclaration => "import_declaration",
            Self::ImportPath => "import_path",

            // Annotations
            Self::Annotation => "annotation",
            Self::Annotations => "annotations",
            Self::AnnotationDeclaration => "annotation_declaration",

            // JSON
            Self::JsonValue => "json_value",
            Self::JsonNumber => "json_number",
            Self::JsonString => "json_string",
            Self::JsonArray => "json_array",
            Self::JsonObject => "json_object",
            Self::JsonObjectPair => "json_object_pair",

            // Error
            Self::Error => "ERROR",
        }
    }

    // Comments and Documentation
    pub fn is_comment(n: &Node) -> bool {
        n.kind() == Self::Comment.as_str()
    }

    pub fn is_docstring(n: &Node) -> bool {
        n.kind() == Self::Docstring.as_str()
    }

    // Identifiers and Names
    pub fn is_identifier(n: &Node) -> bool {
        n.kind() == Self::Identifier.as_str()
    }

    pub fn is_module_definition(n: &Node) -> bool {
        n.kind() == Self::ModuleDefinition.as_str()
    }

    pub fn is_scoped_name(n: &Node) -> bool {
        n.kind() == Self::ScopedName.as_str()
    }

    // Type System
    pub fn is_type_name(n: &Node) -> bool {
        n.kind() == Self::TypeName.as_str()
    }

    pub fn is_type_parameters(n: &Node) -> bool {
        n.kind() == Self::TypeParameters.as_str()
    }

    pub fn is_type_expression(n: &Node) -> bool {
        n.kind() == Self::TypeExpression.as_str()
    }

    pub fn is_type_arguments(n: &Node) -> bool {
        n.kind() == Self::TypeArguments.as_str()
    }

    pub fn is_primitive_type(n: &Node) -> bool {
        n.kind() == Self::PrimitiveType.as_str()
    }

    // Structs
    pub fn is_struct_definition(n: &Node) -> bool {
        n.kind() == Self::StructDefinition.as_str()
    }

    pub fn is_field_block(n: &Node) -> bool {
        n.kind() == Self::FieldBlock.as_str()
    }

    pub fn is_field(n: &Node) -> bool {
        n.kind() == Self::Field.as_str()
    }

    // Unions
    pub fn is_union_definition(n: &Node) -> bool {
        n.kind() == Self::UnionDefinition.as_str()
    }

    // Type Definitions
    pub fn is_type_definition(n: &Node) -> bool {
        n.kind() == Self::TypeDefinition.as_str()
    }

    pub fn is_newtype_definition(n: &Node) -> bool {
        n.kind() == Self::NewtypeDefinition.as_str()
    }

    // Module System
    pub fn is_module(n: &Node) -> bool {
        n.kind() == Self::Module.as_str()
    }

    pub fn is_module_body(n: &Node) -> bool {
        n.kind() == Self::ModuleBody.as_str()
    }

    pub fn is_import_declaration(n: &Node) -> bool {
        n.kind() == Self::ImportDeclaration.as_str()
    }

    pub fn is_import_path(n: &Node) -> bool {
        n.kind() == Self::ImportPath.as_str()
    }

    // Annotations
    pub fn is_annotation(n: &Node) -> bool {
        n.kind() == Self::Annotation.as_str()
    }

    pub fn is_annotations(n: &Node) -> bool {
        n.kind() == Self::Annotations.as_str()
    }

    pub fn is_annotation_declaration(n: &Node) -> bool {
        n.kind() == Self::AnnotationDeclaration.as_str()
    }

    // JSON
    pub fn is_json_value(n: &Node) -> bool {
        n.kind() == Self::JsonValue.as_str()
    }

    pub fn is_json_number(n: &Node) -> bool {
        n.kind() == Self::JsonNumber.as_str()
    }

    pub fn is_json_string(n: &Node) -> bool {
        n.kind() == Self::JsonString.as_str()
    }

    pub fn is_json_array(n: &Node) -> bool {
        n.kind() == Self::JsonArray.as_str()
    }

    pub fn is_json_object(n: &Node) -> bool {
        n.kind() == Self::JsonObject.as_str()
    }

    pub fn is_json_object_pair(n: &Node) -> bool {
        n.kind() == Self::JsonObjectPair.as_str()
    }

    // Error
    pub fn is_error(n: &Node) -> bool {
        n.kind() == Self::Error.as_str()
    }
}
