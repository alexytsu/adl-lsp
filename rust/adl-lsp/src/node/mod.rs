mod kind;

pub use kind::NodeKind;
use tracing::warn;
use tree_sitter::{Node, TreeCursor};

/// Type-safe helpers to perform operations on known-node types
#[allow(dead_code)]
pub enum AdlNode<'a> {
    Import(AdlImportDeclaration<'a>),
    ScopedName(AdlScopedName<'a>),
    ModuleDefinition(AdlModuleDefinition<'a>),
    ModuleBody(AdlModuleBody<'a>),
    TypeDefinition(AdlTypeDefinition<'a>),
    NewtypeDefinition(AdlNewtypeDefinition<'a>),
    StructDefinition(AdlStructDefinition<'a>),
    UnionDefinition(AdlUnionDefinition<'a>),
    Field(AdlField<'a>),
    AnnotationDeclaration(AdlAnnotationDeclaration<'a>),
}

/// A module definition provides a namespace for types and functions that are defined within it
///
/// ### Grammar definition
/// ```javascript
/// module_definition: ($) => seq(
///   optional($.definition_preamble),
///   "module",
///   $.scoped_name,
///   $.module_body
/// ),
/// ```
#[derive(Debug)]
pub struct AdlModuleDefinition<'a> {
    node: Node<'a>,
}

impl<'a> AdlModuleDefinition<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_module_definition(&node) {
            Some(AdlModuleDefinition { node })
        } else {
            warn!("expected a module definition node but got {}", node.kind());
            None
        }
    }

    pub fn module_name<'b>(&self, content: &'b [u8]) -> &'b str {
        let module_or_preamble = self
            .node
            .child(0)
            .expect("module_definition should have children");
        let scoped_name = if NodeKind::is_definition_preamble(&module_or_preamble) {
            self.node.child(2).expect("expected scoped_name")
        } else {
            self.node.child(1).expect("expected scoped_name")
        };
        scoped_name.utf8_text(content).expect("utf-8 parse error")
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.node.walk();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// An import declaration appears at the top of a module and can either be a fully qualified name or a star-import.
///
/// ### Grammar definition
/// ```javascript
/// import_declaration: ($) => seq("import", $.import_path, ";"),
/// import_path: ($) => seq($.scoped_name, optional(".*")),
/// ```
#[derive(Debug, Clone)]
pub enum AdlImportDeclaration<'a> {
    FullyQualified(AdlScopedName<'a>),
    StarImport(AdlScopedName<'a>),
}

impl<'a> AdlImportDeclaration<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_import_declaration(&node) {
            if Self::is_star_import(&node) {
                Some(AdlImportDeclaration::StarImport(AdlScopedName { node }))
            } else {
                Some(AdlImportDeclaration::FullyQualified(AdlScopedName { node }))
            }
        } else {
            None
        }
    }

    pub fn is_star_import(node: &Node<'a>) -> bool {
        node.child(1)
            .expect("expected import_path")
            .child(1)
            .is_some() // either scoped name or scoped_name.*
    }

    pub fn module_name<'b>(&self, content: &'b [u8]) -> &'b str {
        match self {
            AdlImportDeclaration::FullyQualified(scoped_name) => {
                let full_name = scoped_name
                    .node
                    .child(1)
                    .expect("expected import_path")
                    .utf8_text(content)
                    .expect("utf-8 parse error");
                // Remove the last part (type name) from fully qualified names
                // e.g., "mod.foo.Type" -> "mod.foo"
                if let Some(last_dot_pos) = full_name.rfind('.') {
                    &full_name[..last_dot_pos]
                } else {
                    full_name
                }
            }
            AdlImportDeclaration::StarImport(scoped_name) => scoped_name
                .node
                .child(1)
                .expect("expected import_path")
                .child(0) // drop the .*
                .expect("expected scoped_name")
                .utf8_text(content)
                .expect("utf-8 parse error"),
        }
    }

    pub fn imported_type_name<'b>(&self, content: &'b [u8]) -> Option<&'b str> {
        match self {
            AdlImportDeclaration::FullyQualified(scoped_name) => scoped_name
                .node
                .child(1)
                .expect("expected import_path")
                .utf8_text(content)
                .expect("utf-8 parse error")
                .split('.')
                .last(),
            AdlImportDeclaration::StarImport(_) => None,
        }
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let node = match self {
            AdlImportDeclaration::FullyQualified(scoped_name) => &scoped_name.node,
            AdlImportDeclaration::StarImport(scoped_name) => &scoped_name.node,
        };
        let mut cursor = node.walk();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// A scoped name is a sequence of identifiers separated by dots
/// It is used to define module paths and fully-qualified type names in ADL
///
/// ### Grammar definition
/// ```javascript
/// scoped_name: ($) => seq($.identifier, repeat(seq(".", $.identifier))),
/// ```
#[derive(Debug, Clone)]
pub struct AdlScopedName<'a> {
    node: Node<'a>,
}

#[cfg(test)]
mod test {
    use lsp_types::Url;

    use super::AdlImportDeclaration;
    use crate::node::NodeKind;
    use crate::parser::{AdlParser, tree::Tree};

    const SAMPLE_ADL: &str = r#"
    module foo {
        import foo.bar.Baz;
        import foo.char.*;
    };
    "#;

    #[test]
    pub fn test_module_definition_basic() {
        let mut parser = AdlParser::new();
        let uri: Url = "file://input/message.adl".parse().unwrap();
        let tree = parser.parse(uri, SAMPLE_ADL).unwrap();
        let module_definition = tree
            .find_first_node(NodeKind::is_module_definition)
            .expect("expected module definition");
        let module_name = super::AdlModuleDefinition::try_new(module_definition)
            .expect("expected module definition")
            .module_name(SAMPLE_ADL.as_bytes());
        assert_eq!(module_name, "foo");
    }

    #[test]
    pub fn test_import_declaration_basic() {
        let mut parser = AdlParser::new();
        let uri: Url = "file://input/message.adl".parse().unwrap();
        let tree = parser.parse(uri, SAMPLE_ADL).unwrap();
        let import_declarations = tree
            .find_all_nodes(NodeKind::is_import_declaration)
            .iter()
            .filter_map(|n| AdlImportDeclaration::try_new(*n))
            .collect::<Vec<_>>();

        assert_eq!(import_declarations.len(), 2);
        assert!(matches!(
            import_declarations[0],
            AdlImportDeclaration::FullyQualified(_)
        ));
        assert!(matches!(
            import_declarations[1],
            AdlImportDeclaration::StarImport(_)
        ));

        assert_eq!(
            import_declarations[0].module_name(SAMPLE_ADL.as_bytes()),
            "foo.bar"
        );
        assert_eq!(
            import_declarations[1].module_name(SAMPLE_ADL.as_bytes()),
            "foo.char"
        );
        assert_eq!(
            import_declarations[0].imported_type_name(SAMPLE_ADL.as_bytes()),
            Some("Baz")
        );
        assert_eq!(
            import_declarations[1].imported_type_name(SAMPLE_ADL.as_bytes()),
            None
        );
    }
}

/// A module body contains imports, type definitions and annotation declarations
///
/// ## Grammar Definition
/// ```javascript
/// module_body: ($) =>
///   seq(
///     "{",
///     repeat(
///       choice(
///         $.import_declaration,
///         $.annotation_declaration,
///         $.type_definition,
///         $.newtype_definition,
///         $.struct_definition,
///         $.union_definition
///       )
///     ),
///     "}"
/// )
/// ```
#[derive(Debug)]
pub struct AdlModuleBody<'a> {
    node: Node<'a>,
}

impl<'a> AdlModuleBody<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_module_body(&node) {
            Some(AdlModuleBody { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }
}

/// A type definition defines a new type
///
/// ### Grammar definition
/// ```javascript
/// type_definition: ($) =>
///   seq(
///     optional($.definition_preamble),
///     "type",
///     $.type_name,
///     optional($.type_parameters),
///     "=",
///     $.type_expression,
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlTypeDefinition<'a> {
    node: Node<'a>,
}

impl<'a> AdlTypeDefinition<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_type_definition(&node) {
            Some(AdlTypeDefinition { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// A newtype definition defines a new type with a default value
///
/// ### Grammar definition
/// ```javascript
/// newtype_definition: ($) =>
///   seq(
///     optional($.definition_preamble),
///     "newtype",
///     $.type_name,
///     optional($.type_parameters),
///     "=",
///     $.type_expression,
///     optional(seq("=", $.json_value)),
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlNewtypeDefinition<'a> {
    node: Node<'a>,
}

impl<'a> AdlNewtypeDefinition<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_newtype_definition(&node) {
            Some(AdlNewtypeDefinition { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// A struct definition defines a new type with named fields
///
/// ### Grammar definition
/// ```javascript
/// struct_definition: ($) =>
///   seq(
///     optional($.definition_preamble),
///     "struct",
///     $.type_name,
///     optional($.type_parameters),
///     $.field_block,
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlStructDefinition<'a> {
    node: Node<'a>,
}

impl<'a> AdlStructDefinition<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_struct_definition(&node) {
            Some(AdlStructDefinition { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// A union definition defines a new type with named fields
///
/// ### Grammar definition
/// ```javascript
/// union_definition: ($) =>
///   seq(
///     optional($.definition_preamble),
///     "union",
///     $.type_name,
///     optional($.type_parameters),
///     $.field_block,
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlUnionDefinition<'a> {
    node: Node<'a>,
}

impl<'a> AdlUnionDefinition<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_union_definition(&node) {
            Some(AdlUnionDefinition { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// A field defines a named field in a struct or union
///
/// ### Grammar definition
/// ```javascript
/// field: ($) =>
///   seq(
///     optional($.definition_preamble),
///     $.type_expression,
///     $.identifier,
///     optional(seq("=", $.json_value)),
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlField<'a> {
    node: Node<'a>,
}

impl<'a> AdlField<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_field(&node) {
            Some(AdlField { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}

/// An annotation declaration defines a new annotation
///
/// ### Grammar definition
/// ```javascript
/// annotation_declaration: ($) =>
///   seq(
///     "annotation",
///     seq($.scoped_name, repeat(seq("::", $.field_reference))),
///     $.scoped_name,
///     $.json_value,
///     optional(";")
///   )
/// ```
#[derive(Debug)]
pub struct AdlAnnotationDeclaration<'a> {
    node: Node<'a>,
}

impl<'a> AdlAnnotationDeclaration<'a> {
    pub fn try_new(node: Node<'a>) -> Option<Self> {
        if NodeKind::is_annotation_declaration(&node) {
            Some(AdlAnnotationDeclaration { node })
        } else {
            None
        }
    }

    pub fn cursor(&self) -> TreeCursor {
        self.node.walk()
    }

    pub fn is_missing_semicolon(&self) -> bool {
        let mut cursor = self.cursor();
        let last_child = cursor.goto_last_child();
        last_child && cursor.node().kind() != ";"
    }
}
