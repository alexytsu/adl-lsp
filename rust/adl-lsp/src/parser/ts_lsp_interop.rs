use async_lsp::lsp_types::{Position, Range};
use tree_sitter::{Point, Range as TsRange};

pub fn ts_to_lsp_position(p: &Point) -> Position {
    Position {
        line: p.row as u32,
        character: p.column as u32,
    }
}

pub fn ts_to_lsp_range(r: &TsRange) -> Range {
    Range {
        start: ts_to_lsp_position(&r.start_point),
        end: ts_to_lsp_position(&r.end_point),
    }
}

#[allow(unused)]
pub fn lsp_to_ts_point(p: &Position) -> Point {
    Point {
        row: p.line as usize,
        column: p.character as usize,
    }
}
