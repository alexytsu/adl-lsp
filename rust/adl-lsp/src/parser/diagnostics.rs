use async_lsp::lsp_types::{Diagnostic, DiagnosticSeverity, Range};
use tracing::debug;

use crate::node::NodeKind;
use crate::parser::tree::Tree;
use crate::parser::ts_lsp_interop::ts_to_lsp_position;

use super::ParsedTree;

impl ParsedTree {
    pub fn collect_diagnostics(&self, content: &str) -> Vec<Diagnostic> {
        if content.trim().is_empty() {
            return vec![Diagnostic {
                severity: Some(DiagnosticSeverity::WARNING),
                message: "Empty file".to_string(),
                ..Default::default()
            }];
        }

        let mut diagnostics: Vec<Diagnostic> = Vec::new();
        // first collect parse errors
        self.collect_parse_diagnostics(&mut diagnostics);
        self.collect_parse_diagnostics_missing(&mut diagnostics);

        // then collect custom semantic errors
        if let Some(import_diagnostics) = self.collect_import_diagnostics() {
            diagnostics.extend(import_diagnostics);
        }

        debug!("collected diagnostics: {:?}", diagnostics);
        diagnostics
    }

    pub fn collect_parse_diagnostics(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_error)
                .into_iter()
                .map(|n| Diagnostic {
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                    severity: Some(DiagnosticSeverity::ERROR),
                    message: "Syntax error".to_string(),
                    ..Default::default()
                }),
        );
    }

    pub fn collect_parse_diagnostics_missing(&self, diagnostics: &mut Vec<Diagnostic>) {
        diagnostics.extend(
            self.find_all_nodes(NodeKind::is_missing)
                .into_iter()
                .map(|n| Diagnostic {
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                    severity: Some(DiagnosticSeverity::ERROR),
                    message: "Missing token '".to_string() + n.kind() + "'",
                    ..Default::default()
                }),
        );
    }

    pub fn collect_import_diagnostics(&self) -> Option<Vec<Diagnostic>> {
        let imports = self.find_all_nodes(NodeKind::is_import_declaration);
        let module_body_children = self.find_first_node(NodeKind::is_module_body)?;
        let mut cursor = module_body_children.walk();
        cursor.goto_first_child(); // opening module brace

        let mut first_non_import = None;
        while cursor.goto_next_sibling() {
            let node = cursor.node();
            if !NodeKind::is_import_declaration(&node)
                && !NodeKind::is_docstring(&node)
                && !NodeKind::is_comment(&node)
            {
                first_non_import = Some(node);
                break;
            }
            cursor.goto_next_sibling();
        }

        let first_non_import = first_non_import?;

        debug!("first_non_import: {:?}", first_non_import.kind());

        // TODO: attempt to resolve the imports and report errors for invalid imports

        let diagnostics = imports
            .iter()
            .filter_map(|node| {
                // imports should only be at the top of a module
                if node.start_position() > first_non_import.start_position() {
                    Some(Diagnostic {
                        range: Range {
                            start: ts_to_lsp_position(&node.start_position()),
                            end: ts_to_lsp_position(&node.end_position()),
                        },
                        message: "Imports must be at the top of a module".to_string(),
                        severity: Some(DiagnosticSeverity::ERROR),
                        code: None,
                        code_description: None,
                        source: Some("adl-lsp".to_string()),
                        related_information: None,
                        tags: None,
                        data: None,
                    })
                } else {
                    None
                }
            })
            .collect();

        Some(diagnostics)
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::AdlParser;

    #[test]
    fn test_collect_parse_error() {
        let url: Url = "file://foo/error.adl".parse().unwrap();
        let contents = include_str!("input/error.adl");

        let parsed = AdlParser::new().parse(url.clone(), &contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(contents));
    }

    #[test]
    fn test_collect_import_error() {
        let url: Url = "file://foo/importerror.adl".parse().unwrap();
        let contents = include_str!("input/importerror.adl");

        let parsed = AdlParser::new().parse(url.clone(), &contents);
        assert!(parsed.is_some());
        assert_yaml_snapshot!(parsed.unwrap().collect_diagnostics(&contents));
    }
}
