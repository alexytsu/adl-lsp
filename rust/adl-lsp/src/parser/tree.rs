use async_lsp::lsp_types::Position;
use tree_sitter::{Node, TreeCursor};

use crate::parser::ts_lsp_interop as ts_lsp;
use crate::{node::NodeKind, parser::ParsedTree};

/// Basic tree traversal methods for working with the tree-sitter parsed tree
pub trait Tree {
    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool;
    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>>;
    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>>;
    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node>;
    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>>;
    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>>;
}

impl Tree for ParsedTree {
    fn walk_and_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
        early: bool,
    ) -> Vec<Node<'a>> {
        let mut v = vec![];

        loop {
            let node = cursor.node();

            if f(&node) {
                v.push(node);
                if early {
                    break;
                }
            }

            if cursor.goto_first_child() {
                v.extend(Self::walk_and_filter(cursor, f, early));
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }

        v
    }

    fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool {
        loop {
            let node = cursor.node();
            if node.id() == nid {
                return true;
            }
            if cursor.goto_first_child() {
                if Self::advance_cursor_to(cursor, nid) {
                    return true;
                }
                cursor.goto_parent();
            }
            if !cursor.goto_next_sibling() {
                return false;
            }
        }
    }

    fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>> {
        let pos = ts_lsp::lsp_to_ts_point(pos);
        self.tree.root_node().descendant_for_point_range(pos, pos)
    }

    fn find_all_nodes(&self, f: fn(&Node) -> bool) -> Vec<Node> {
        self.find_all_nodes_from(self.tree.root_node(), f)
    }

    fn find_all_nodes_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, false)
    }

    fn find_first_node<'a>(&'a self, f: fn(&Node) -> bool) -> Option<Node<'a>> {
        self.find_node_from(self.tree.root_node(), f).first().copied()
    }

    fn find_node_from<'a>(&self, n: Node<'a>, f: fn(&Node) -> bool) -> Vec<Node<'a>> {
        let mut cursor = n.walk();
        Self::walk_and_filter(&mut cursor, f, true)
    }
}

impl ParsedTree {
    pub fn get_identifier_at<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<(&'a str, Node<'a>)> {
        self.get_node_at_position(pos)
            .filter(NodeKind::is_identifier)
            .map(|n| (n.utf8_text(content.as_ref()).expect("utf-8 parse error"), n))
    }
}
