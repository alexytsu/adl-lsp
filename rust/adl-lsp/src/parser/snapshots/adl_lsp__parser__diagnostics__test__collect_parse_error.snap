---
source: src/parser/diagnostics.rs
expression: parsed.unwrap().collect_parse_diagnostics()
---
- range:
    start:
      line: 3
      character: 16
    end:
      line: 7
      character: 14
  severity: 1
  source: adl-vscode
  message: Syntax error
- range:
    start:
      line: 8
      character: 4
    end:
      line: 8
      character: 6
  severity: 1
  source: adl-vscode
  message: Syntax error
